lib/links/parser.rb

```diff
@@ -0,0 +1,65 @@
+require 'httparty'
+
+class Parser
+
+  @@elements = { a: ['href'],
+                img: ['src'],
+                script: ['src'],
+                frame: ['src'],
+                iframe: ['src'],
+                link: ['href'],
+                source: ['src', 'srcset'],
+                track: ['src'] }
+
+  attr_reader :url
+  #attr_accessor :doc
+
+  def initialize(url)
+    @url = url
+  end
+
+  def run
+    all_links = []
+    @doc ||= get_html
+
+    @@elements.each do |elem, attr_list|
+      elems = get_elems(elem)
+
+      attr_list.each do |attr|
+        links = extract_links(elems, attr)
+        all_links = (all_links << links).flatten
+      end
+    end
+
+    all_links.uniq
+  end
+
+  def get_html
+    Nokogiri::HTML(HTTParty.get(url))
```

Comment:
How about saving the intermediate result with `@doc ||= ...` so you can call this multiple times? And if you do that, a better name would be `doc_html` or something, since it's not always `get`ting the contents.

Reply:
The idea was that I would use this method as I wish. For example, if I recursively went inside other links, I would do `@doc = get_html` as opposed to `@doc ||= get_html` in this situation. Or am I getting this all wrong?

Reply:
It's just for convenience. A common pattern to access the variable without caring for loading is to do the following:

```ruby

private

def doc
  @_doc ||= get_html
end

def get_html
  # what you were already doing
end
```


And now you can do the following:

```ruby
def run
  doc.map do |link|
    # ...
    link unless doc.include?(link) # just an example to show how you can use the variable twice
  end
end
```

In the scenario above you can call `doc` twice and it will only make the HTTP request once. It's a common pattern.

Reply:
Ok, now I get it. Thanks!